File: apps/api/app/api/prompts.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.api.deps import (
    assert_brand_visible,
    assert_prompt_visible,
    current_principal,
    get_db,
    require_write,
    visible_brand_ids,
)
from app.core.security import Principal
from app.schemas.prompt import PromptCreate, PromptListOut, PromptOut, PromptUpdate
from app.services import prompts as prompt_svc

router = APIRouter(prefix="/v1/prompts", tags=["prompts"])


@router.get("", response_model=PromptListOut)
def list_prompts(
    brand_id: Optional[int] = Query(None),
    active_only: bool = Query(False),
    db: Session = Depends(get_db),
    principal: Principal = Depends(current_principal),
):
    if brand_id is not None:
        assert_brand_visible(db, principal, brand_id)
    else:
        # 不带 brand_id 时客户不能拿到全部 —— 收敛到他自己的品牌集合
        allowed = visible_brand_ids(db, principal)
        if allowed is not None:
            items = [
                pr
                for bid in allowed
                for pr in prompt_svc.list_prompts(db, brand_id=bid, active_only=active_only)
            ]
            return PromptListOut(
                items=[PromptOut.model_validate(x) for x in items], total=len(items)
            )
    items = prompt_svc.list_prompts(db, brand_id=brand_id, active_only=active_only)
    return PromptListOut(
        items=[PromptOut.model_validate(p) for p in items],
        total=prompt_svc.count_prompts(db, brand_id=brand_id),
    )


@router.post("", response_model=PromptOut, status_code=status.HTTP_201_CREATED)
def create_prompt(
    body: PromptCreate,
    db: Session = Depends(get_db),
    _: Principal = Depends(require_write),
):
    p = prompt_svc.create_prompt(db, body)
    return PromptOut.model_validate(p)


@router.get("/{prompt_id}", response_model=PromptOut)
def get_prompt(
    prompt_id: int,
    db: Session = Depends(get_db),
    principal: Principal = Depends(current_principal),
):
    assert_prompt_visible(db, principal, prompt_id)
    p = prompt_svc.get_prompt_or_404(db, prompt_id)
    return PromptOut.model_validate(p)


@router.patch("/{prompt_id}", response_model=PromptOut)
def update_prompt(
    prompt_id: int,
    body: PromptUpdate,
    db: Session = Depends(get_db),
    _: Principal = Depends(require_write),
):
    p = prompt_svc.update_prompt(db, prompt_id, body)
    return PromptOut.model_validate(p)


@router.delete("/{prompt_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_prompt(
    prompt_id: int,
    db: Session = Depends(get_db),
    _: Principal = Depends(require_write),
):
    prompt_svc.delete_prompt(db, prompt_id)
    return None

File: apps/web/src/lib/api/prompts.ts
/**
 * 提问词。字段以 `docs/API.md` §6 为准。
 */
import type { Prompt } from '../types'
import { apiFetch } from './client'

export interface PromptListResult {
  items: Prompt[]
  total: number
}

/**
 * 某品牌的提问词。
 *
 * `activeOnly` 默认 false —— 管理界面要看见**停用的那些**，否则用户会以为
 * 它们被删了。发起 run 时后端只取 `is_active=true`（`services/tasks.create_run`），
 * 那是另一回事。
 *
 * **`category` 没有服务端过滤参数**（API.md §6）—— 拿回来自己归类。
 */
export async function listPrompts(params: {
  brandId: number
  activeOnly?: boolean
}): Promise<PromptListResult> {
  return apiFetch<PromptListResult>('/v1/prompts', {
    query: { brand_id: params.brandId, active_only: params.activeOnly },
  })
}

export interface PromptCreateInput {
  brand_id: number
  text: string
  category?: string | null
  is_active?: boolean
}

export async function createPrompt(input: PromptCreateInput): Promise<Prompt> {
  return apiFetch<Prompt>('/v1/prompts', { method: 'POST', body: input })
}

/** 改提问词。PATCH 是增量 —— 只传要改的字段。 */
export async function updatePrompt(
  promptId: number,
  input: { text?: string; category?: string | null; is_active?: boolean },
): Promise<Prompt> {
  return apiFetch<Prompt>(`/v1/prompts/${promptId}`, { method: 'PATCH', body: input })
}

/**
 * 删提问词。**这会改写历史，不只是「以后不问了」。**
 *
 * `crawl_jobs.prompt_id` 是 `ON DELETE CASCADE`，所以删一条提问词会连带删掉
 * 它的全部 job → response → mention。后果是：
 *
 *   · 历史 run 的**快照行还在**（`run_prompts.prompt_id` 是普通整数、
 *     没有外键，这是刻意的），所以矩阵里那一行不会消失
 *   · 但那一行的 `n` 掉到 0，变成「不可算」
 *   · 而那次 run 的 `n_valid` 少了一截 —— **总提及率会当场变**
 *
 * 「以后不再问这条」应该用 `updatePrompt(id, { is_active: false })`，
 * 它只影响未来的 run，一个历史数字都不动。
 */
export async function deletePrompt(promptId: number): Promise<void> {
  await apiFetch(`/v1/prompts/${promptId}`, { method: 'DELETE' })
}
